planning/autoware_freespace_planning_algorithms/src/astar_search.cpp

```diff
@@ -333,46 +333,60 @@ void AstarSearch::setPath(const AstarNode & goal_node)
   header.stamp = rclcpp::Clock(RCL_ROS_TIME).now();
   header.frame_id = costmap_.header.frame_id;
 
-  waypoints_.header = header;
-  waypoints_.waypoints.clear();
-
   // From the goal node to the start node
   const AstarNode * node = &goal_node;
 
+  std::vector<PlannerWaypoint> waypoints;
+
   geometry_msgs::msg::PoseStamped pose;
   pose.header = header;
 
-  const auto interpolate = [this, &pose](const AstarNode & node) {
+  const auto interpolate = [this, &waypoints, &pose](const AstarNode & node) {
     if (node.parent == nullptr || !astar_param_.adapt_expansion_distance) return;
     const auto parent_pose = node2pose(*node.parent);
     const double distance_2d = calcDistance2d(node2pose(node), parent_pose);
     const int n = static_cast<int>(distance_2d / min_expansion_dist_);
     for (int i = 1; i < n; ++i) {
-      const double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
+      double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
+      if (is_backward_search_) dist *= -1.0;
       const double steering = node.steering_index * steering_resolution_;
       const auto local_pose = kinematic_bicycle_model::getPose(
         parent_pose, collision_vehicle_shape_.base_length, steering, dist);
       pose.pose = local2global(costmap_, local_pose);
-      waypoints_.waypoints.push_back({pose, node.is_back});
+      waypoints.push_back({pose, node.is_back});
     }
   };
 
   // push astar nodes poses
   while (node != nullptr) {
     pose.pose = local2global(costmap_, node2pose(*node));
-    waypoints_.waypoints.push_back({pose, node->is_back});
+    waypoints.push_back({pose, node->is_back});
     interpolate(*node);
     // To the next node
     node = node->parent;
   }
 
-  // Reverse the vector to be start to goal order
-  std::reverse(waypoints_.waypoints.begin(), waypoints_.waypoints.end());
+  if (waypoints.size() > 1) waypoints.back().is_back = waypoints.rbegin()[1].is_back;
+
+  if (!is_backward_search_) {
+    // Reverse the vector to be start to goal order
+    std::reverse(waypoints.begin(), waypoints.end());
+  }
 
-  // Update first point direction
-  if (waypoints_.waypoints.size() > 1) {
-    waypoints_.waypoints.at(0).is_back = waypoints_.waypoints.at(1).is_back;
+  waypoints_.header = header;
+  waypoints_.waypoints.clear();
+  auto it = waypoints.begin();
+  for (; it < waypoints.end() - 1; ++it) {
+    auto next_it = it + 1;
+    waypoints_.waypoints.push_back(*it);
+    if (it->is_back == next_it->is_back) continue;
+    if (is_backward_search_) {
+      waypoints_.waypoints.push_back({next_it->pose, it->is_back});
+    } else {
+      waypoints_.waypoints.push_back({it->pose, next_it->is_back});
+    }
   }
```

Comment:
```suggestion
  waypoints_.waypoints.reserve(waypoints.size());
  
  for (size_t i = 0; i < waypoints.size() - 1; ++i) {
    const auto& current = waypoints[i];
    const auto& next = waypoints[i + 1];

    waypoints_.waypoints.push_back(current);

    if (current.is_back != next.is_back) {
      waypoints_.waypoints.push_back({
        is_backward_search_ ? next.pose : current.pose,
        is_backward_search_ ? current.is_back : next.is_back
      });
    }
  }
```

Maybe something like this to keep it more readable.